internal/types/compare.go

```diff
@@ -254,3 +258,103 @@ func compareNumbers(a float64, b int64) CompareResult {
 
 	return CompareResult(bigA.Cmp(bigB))
 }
+
+// CompareArrays compares indices of a filter array according to indices of a document array;
+// returns Equal when a document array contains another array(subarray) that equals filter array.
+func CompareArrays(filterArr, docArr *Array) CompareResult {
+	if docArr.Len() == 0 && filterArr.Len() == 0 {
+		return Equal
+	}
+	if filterArr.Len() == 0 {
+		return Incomparable
+	}
+
+	EntireArrayResult, subArrayEquality := Incomparable, Incomparable
+
+	for i := 0; i < docArr.Len(); i++ {
+		arrValue := must.NotFail(docArr.Get(i))
+		switch arrValue := arrValue.(type) {
+		case *Array:
+			filterArrValue := must.NotFail(filterArr.Get(i))
+			switch filterArrValue := filterArrValue.(type) {
+			case *Array:
+				res := CompareArrays(filterArrValue, arrValue)
+				res = handleSubArrayComparingResult(i, &res, &EntireArrayResult, &subArrayEquality)
+				continue
+
+			default:
+				res := CompareArrays(filterArr, arrValue)
+				res = handleSubArrayComparingResult(i, &res, &EntireArrayResult, &subArrayEquality)
+			}
+			continue
+
+		// TODO: case Document
+		// case *Document
+
+		default:
+			if i+1 > filterArr.Len() {
+				if EntireArrayResult == Equal {
+					EntireArrayResult = Incomparable
+				}
+				continue // looking for next element is array that might fit filter query
+			}
+
+			filterValue := must.NotFail(filterArr.Get(i))
+			switch filterValue := filterValue.(type) {
+			case *Array, *Document:
+				if EntireArrayResult == Equal {
+					EntireArrayResult = Incomparable
+				}
+				continue
+			default:
+				res := CompareOrder(arrValue, filterValue, Ascending)
+				if EntireArrayResult == Incomparable && i == 0 { // set first non-Incomparable result
+					EntireArrayResult = res
+				}
+
+				if EntireArrayResult != res {
+					EntireArrayResult = Incomparable
```

Comment:
here and there are also early returns. Isn't it?

Reply:
looking for next element/value is array that might fit filter array.